### ISSUE TITLE
open to a PR for for a container query plugin?

### ISSUE THREAD

**Issue by mreinstein** [snabbdom/snabbdom#925]
it would be really nice to have some way to be able to do container-queries (being able to set classes on a DOM element based on some criteria like element width.)

Happy to try to write one if one doesn't already exist, just don't want to re-invent the wheel here.

**Comment by mreinstein** [comment]
I've put together a proof-of-concept, here's a demo of it in action:

https://test-snabbdom-container-query.nekoflux.vercel.app/

If you inspect the `<main>` element and resize the window, you'll see different classes (`BP1`, `BP2`, `BP3`) applied to the element. This looks like a `@media` query but it's actually specified at the component level!

```js
h('main', {
    attrs: {
        'data-breakpoints': '{"BP1":400,"BP2":800,"BP3":1200}'
    },
  }, `Just some test string`)
```

Is this something we'd be interested in adding as another optional module to snabbdom? I'm happy to put this into a PR if so.

The only production issue with this module is it assumes presence of the `ResizeObserver` API.  Support is overall quite good but I think we'd want to ponyfill that capability as needed, which would add a little complexity to the implementation, but nothing crazy I think.


my current implementation:

snabbdom-containerquery.js:
```js
// inspired by https://philipwalton.com/articles/responsive-components-a-solution-to-the-container-queries-problem/

// currently, it's more performant to have 1 resizeObserver observe multiple elements rather than 1 per element.
// see https://groups.google.com/a/chromium.org/g/blink-dev/c/z6ienONUb5A/m/F5-VcUZtBAAJ
const observer = new ResizeObserver(function (entries) {
	for (const entry of entries) {
		const breakpoints = JSON.parse(entry.target.dataset.breakpoints);
		let prev;

		for (const breakpoint of Object.keys(breakpoints)) {
			const minWidth = breakpoints[breakpoint];
	        if (entry.contentRect.width >= minWidth) {
	        	entry.target.classList.add(breakpoint);

	        	// exclusively match only a single breakpoint, so remove any other breakpoint
	        	// classes that might have previously matched.
	        	if (prev)
	        		entry.target.classList.remove(prev);

	        } else {
	        	entry.target.classList.remove(breakpoint);
	        }

	        prev = breakpoint;
		}
	}
});


function update (oldVnode, vnode) {
	if (vnode.elm.dataset.breakpoints) {
		try {
			// ensure that the data attribute parses as valid json before we start observing it
			JSON.parse(vnode.elm.dataset.breakpoints);
			observer.observe(vnode.elm);
		} catch (er) {
			throw new Error('Invalid breakpoints declared. must be valid JSON', er);
		}
	} else {
		observer.unobserve(vnode.elm);
	}
}


function destroy (vnode/*, removeCallback*/) {
	observer.unobserve(vnode.elm);
}


export default {
	create: update,
	update,
	destroy
}

```

**Comment by paldepind** [comment]
I think it would be appropriate to maintain this module as a separate repository/package. That could, if you're interested, be a repository under the Snabbdom GitHub organisation.

**Comment by mreinstein** [comment]
no longer necessary, now that css container queries are approaching widespread availability.

**Comment by paldepind** [comment]
Thank you for taking the time to closing the issue now that it is no longer necessary and thank you for sharing the container query plugin 👍
